Combine Upscale classes

chakshinglee opened the pull request:
Because the core part of all the *Upscale classes are essentially the same (the differences being how we construct the distributed graph and treatment of boundary conditions), this PR collapses them into Upscale. The benefit is to avoid changing all the *Upscale files whenever there is some update to Topology/Coarsener/Solver. 

Brief summary of changes:
1. Extract `ParGraph` class from `utilities.xpp` and rename it to simply `Graph`, which contains distributed vertex edge relation, edge weight, and edge to true edge relation. `Graph` can be constructed by distributing a global graph, or distributed graph. 
2. Use `Graph` as input parameter for `Upscale`, `GraphTopology`, `MixedMatrix`,  coarsener etc.
3. Move the construction of extended aggregates from `GraphTopology` to inside coarsener.
4. Remove `partitioning` in `Graph` because it should not be part of a graph. We had it before because we wanted to preserve the global partitioning when distributing the global graph, so that serial and parallel run will have the same upscaling error (one way to check if the parallel code is correct). Now partitioning is not preserved, parallel and serial run do not have the same results anymore, so some test criteria are rebaselined.  
5. Some renaming of variables (try to distinguish entities and dofs), and extract some steps into subroutines in the computation of targets.

chakshinglee committed:
change naming edge_d_td to edge_trueedge (prepare for multilevel coarsening in which case one edge can have more than one dof, so edge_d_td should be reserved for edge space dof true dof

chakshinglee committed:
more renaming regarding true edge/face

chakshinglee committed:
extract the steps of computing A * A^T

chakshinglee committed:
use the mfem function HypreParMatrix::Threshold() to remove boundary edges in pAggExt_edge_

chakshinglee committed:
rearrange position of functions in GraphTopology.cpp

chakshinglee committed:
move the construction of extended aggregates from GraphTopology to LocalMixedGraphSpectralTargets, this is because the notion of extended aggregates is only used when computing spectral targets, so they don't need to be kept in GraphTopology throughout. Also, in multilevel setting, we need relation between extended aggregates to vertex and edge 'dofs', which do not make much sense in GraphTopology

chakshinglee committed:
change variable names to prepare for ML framework, which distinguish the notion of entity and dof

chakshinglee committed:
extract the process of building 'permutation' matrix that brings dofs from extended aggregates to local processor

chakshinglee committed:
extract the extraction of local edge/vertex dofs in extended aggregates; rename some variables (make them more compact)

chakshinglee committed:
use extended aggregates dofs extraction function in ComputeEdgeTargets; more variable renaming

chakshinglee committed:
save diagonal and off-diagonal blocks of ExtAgg_dof and use the blocks for extraction later (instead of wrap the blocks in each extraction)

chakshinglee committed:
add an enum type to signify dof types (vertex or edge)

chakshinglee committed:
further shorten ComputeVertexTargets - move some minor details from ComputeVertexTargets to the functions that ComputeVertexTargets calls to implement those details

chakshinglee committed:
better documentation

chakshinglee committed:
merge changes from ml-topo and resolve conflicts

chakshinglee committed:
make style

chakshinglee committed:
minor changes

chakshinglee committed:
remove a constructor that shouldn't be added at this point

chakshinglee committed:
add DofHandler class

chakshinglee committed:
extract ParGraph class out from utilities, rename it to Graph

chakshinglee committed:
make style

chakshinglee committed:
Merge branch 'feature/ml-topo' of https://github.com/LLNL/smoothG into feature/ml-coarsen

chakshinglee committed:
Merge pull request #15 from LLNL/feature/dof-handler

add DofHandler class

chakshinglee committed:
rename DofHandler class and files to GraphSpace

chakshinglee committed:
more rename and remove some inappropriate member of GraphSpace

chakshinglee committed:
add a GraphTolopogy constructor using Graph

chakshinglee committed:
fix some 'const' of Graph and GraphTopology (previously those were not const because extended aggregates require hypre, but now extended aggregates are moved away from Graph/Topology)

chakshinglee committed:
add notion of edge weight in Graph class

chakshinglee committed:
rename edge_e_te_ to edge_trueedge_ in Graph class; some other minor changes

chakshinglee committed:
add use of Graph ojects inside GraphUpscale

chakshinglee committed:
make graph_topology_ a const ref instead of a unique_ptr in Mixed_GL_Coarsener

chakshinglee committed:
Merge branch 'master' of https://github.com/LLNL/smoothG into feature/ml-coarsen

chakshinglee committed:
make a parameter struct for spectral coarsener, move SAAMGeParam to utilities.xpp, use parameter struct to pass the parameters, remove ParGraph class (it is not Graph in Graph.xpp)

chakshinglee committed:
minor changes

chakshinglee committed:
add SpectralCoarsenerParameters, not sure if this is the right way to go, but first make two level upscale interface cleaner and then decide later

chakshinglee committed:
minor changes

chakshinglee committed:
resolve conflict from merging nondiagonal-localsolver into ml-coarsen

chakshinglee committed:
remove some read/write function for vectors in edge space (in each processor)

chakshinglee committed:
remove the notion of partition in Graph (because it should not be part of a graph object)

chakshinglee committed:
remove the notion of redistribution for now (to focus on refactor Upscale classes), but will almost certainly reintroduce it later

chakshinglee committed:
fix some function calls due to change of function name

chakshinglee committed:
move read/write vector function from GraphUpscale to Graph, fix some function calls

chakshinglee committed:
more detailed documentation

chakshinglee committed:
fix edge weight distribution

chakshinglee committed:
partition global graph based on number of processors (not the global_partitioning array), so construction of processor to vertex table can be simplified

chakshinglee committed:
extract construction of edge to true edge as an individual function

chakshinglee committed:
minor renaming

chakshinglee committed:
add a Graph constructor that takes distributed graph as input

chakshinglee committed:
remove notion of DistributeWeight in MixedMatrix (distribution will be done in Graph instead)

chakshinglee committed:
add split edge weight (edge weight wrt each vertex) in Graph

chakshinglee committed:
clean up MixedMatrix, takes Graph as input instead of its components

chakshinglee committed:
merge from master

chakshinglee committed:
unified Upscale class

chakshinglee committed:
fix some seg faults

chakshinglee committed:
remove some unnecessary lines

chakshinglee committed:
update PDESampler to use new Upscale. Update fine solver everytime when MakeFineSolver is called

chakshinglee committed:
remove FiniteVolumeUpscale GraphUpscale FiniteVolumeMLMC

chakshinglee committed:
fix vector initialization

chakshinglee committed:
remove redundant Parameter structs

chakshinglee committed:
temporarily removed GraphSpace (will be reintroduced later)

chakshinglee committed:
minor rearrangement and simplification

chakshinglee committed:
make vertex_edge_ not a const ref in GraphTopology. Some minor simplification

chakshinglee committed:
fix seg fault when Graph is constructed from a distributed graph without specifying edge weights (default to unit weight)

chakshinglee committed:
recover a line in SAAMGe call that was accidentally deleted

chakshinglee committed:
fix elimination of Hybridized system (need elimination of assembled global system). rebaseline parallel graph test results because now the global graph is distributed based on number of processors instead of global partitioning.

chakshinglee committed:
rebaseline parallel graph test results again because on tests on travis are performed on 2 processes

chakshinglee committed:
rebaseline parpoweriter test results, but the errors are abnormally high, this needs to be revisited

chakshinglee committed:
rebaseline results of tests concerning parallel graph. Change default number of processors for make test to 2 instead of 4 (to match github)

chakshinglee commented on examples/stest.py:
```diff
@@ -581,7 +644,7 @@ def make_tests():
           "--num-samples", "2"],
          {"fine-mean-l1": 0.54961180496539375,
-          "max-p-error": 0.39696741537314961}]
+          "max-p-error": 0.39869063097389679}]
```
@barker29 In some PR you raised the concern on this mismatch. Does this update match with test result on your machine (up to `1e-4`)? While working on this PR, I realize that the test tolerance on travis is `1e-2` while the default tolerance specified in CMakeLists.txt is `1e-4`, I guess that's why the test is passing on travis but not on our machines.

barker29 replied:
Yes, this matches.

chakshinglee commented on examples/CMakeLists.txt:
```diff
@@ -42,7 +42,7 @@ target_link_libraries(mlmc smoothg ${TPL_LIBRARIES})
 # add tests
 if (NOT DEFINED SMOOTHG_TEST_PROCS)
-    set(SMOOTHG_TEST_PROCS 4)
+    set(SMOOTHG_TEST_PROCS 2)
```
@barker29 The test results on parallel graph now depend on number of processors, so I match this with the value set on travis.

barker29 commented on src/Upscale.hpp:
```diff
@@ -37,6 +37,30 @@ namespace smoothg
 class Upscale : public mfem::Operator
 {
 public:
+    Upscale(const Graph& graph,
```
Can we get a line of documentation for each of these constructors? (also, it might be possible to combine a couple of them by using more default arguments?)

chakshinglee replied:
You are right, I can use more default arguments. In a newer version only one constructor is left

barker29 commented on src/MixedMatrix.hpp:
```diff
@@ -49,26 +49,7 @@ class MixedMatrix
-    MixedMatrix(const mfem::SparseMatrix& vertex_edge,
-                const std::vector<mfem::Vector>& local_weight,
-                const mfem::HypreParMatrix& edge_d_td);
+    MixedMatrix(const Graph& graph, const mfem::SparseMatrix& w_block = SparseIdentity(0));
```
This is a very nice simplification (in general your use of the `Graph` object is very nice and simplifies a lot!)

barker29 commented on src/LocalMixedGraphSpectralTargets.hpp:
```diff
@@ -167,6 +167,11 @@ class LocalMixedGraphSpectralTargets
         const mfem::SparseMatrix* W_local,
         const GraphTopology& graph_topology);
 
+    LocalMixedGraphSpectralTargets(
```
One line of documentation please (this one is actually a bit different from the other constructors, at least the calling sequence).

barker29 commented on src/LocalMixedGraphSpectralTargets.hpp:
```diff
@@ -182,26 +187,45 @@ class LocalMixedGraphSpectralTargets
                  std::vector<mfem::DenseMatrix>& local_vertex_targets,
                  const mfem::Vector& constant_rep);
 private:
+    enum DofType { vdof, edof }; // vertex-based and edge-based dofs
```
I think maybe `vdof` should be `VDOF` - that's the style I usually see for `enum`, but I guess we can make our own decision. What do you think?

chakshinglee replied:
Sounds good to me, this will be updated.

barker29 commented on src/GraphTopology.hpp:
```diff
@@ -78,15 +92,14 @@ class GraphTopology
-                  const mfem::HypreParMatrix& edge_d_td,
-                  const mfem::HypreParMatrix& face_d_td,
-                  const mfem::HypreParMatrix& face_d_td_d);
+                  const mfem::HypreParMatrix& edge_trueedge,
```
These name changes are much better than before.

barker29 commented on examples/stest.py:
```diff
@@ -246,22 +246,43 @@ def make_tests():
-          "finest-p-error": 0.38013398274257243,
-          "finest-u-error": 0.38079825403520218,
+         {"finest-div-error": 0.79929158208615803,
+          "finest-p-error": 0.80196594708999625,
```
As I said in a previous comment, these errors seem really quite high. I want to understand why before I approve this PR.

barker29 commented on examples/stest.py:
```diff
@@ -246,22 +246,43 @@ def make_tests():
+          "finest-u-error": 0.78297677149740286,
           "operator-complexity": 1.016509834901651}]
 
+    tests["parsamplegraph1-coeff-mac"] = \
```
These `-mac` tests are starting to get a bit out of hand. I'd like to have consistent tests that don't depend on platform, but that's probably a harder task and anyway would be a totally different PR.

barker29 commented on src/Graph.cpp:
```diff
@@ -0,0 +1,340 @@
+ *
+ * Copyright (c) 2017,  Lawrence Livermore National Security, LLC.
+ * Produced at the Lawrence Livermore National Laboratory.
+ * LLNL-CODE-XXXXXX. All Rights reserved.
```
This should have the correct release number (in general this should be done before anything is put on GitHub, even in a branch).

barker29 commented on src/Graph.cpp:
```diff
@@ -0,0 +1,340 @@
+ * LLNL-CODE-XXXXXX. All Rights reserved.
+ *
+ * This file is part of smoothG.  See file COPYRIGHT for details.
+ * For more information and source code availability see XXXXX.
```
replace XXXXX

barker29 commented on src/Graph.hpp:
```diff
@@ -0,0 +1,160 @@
+ *
+ * Copyright (c) 2017,  Lawrence Livermore National Security, LLC.
+ * Produced at the Lawrence Livermore National Laboratory.
+ * LLNL-CODE-XXXXXX. All Rights reserved.
```
release number.

barker29 commented on src/Graph.hpp:
```diff
@@ -0,0 +1,160 @@
+    Extract the local submatrix of the global vertex to edge relation table
+    Each vertex belongs to one and only one processor, while some edges are
+    shared by two processors, indicated by the edge to true edge
+    HypreParMatrix edge_e_te
```
```suggestion
    HypreParMatrix edge_trueedge
```

barker29 commented on src/Graph.hpp:
```diff
@@ -0,0 +1,160 @@
+       Based on the partition numbers of vertices in partition_global,
+       distribute a global graph into local subgraphs in each processors, each
+       vertex is local to one and only one processor, while edges can be shared
+       between processors, which is encoded in edge_e_te_.
```
```suggestion
       between processors, which is encoded in edge_trueedge_.
```

barker29 commented on src/Graph.hpp:
```diff
@@ -0,0 +1,160 @@
+          const mfem::SparseMatrix& vertex_edge_global,
+          const mfem::Vector& edge_weight_global = mfem::Vector());
+
+    Graph(const mfem::SparseMatrix& vertex_edge_local,
```
A line of documentation please (it's fine to have thorough documentation for one constructor and short doc for the others, but they should all have something.)

barker29 commented on src/Graph.hpp:
```diff
@@ -0,0 +1,160 @@
+          const mfem::HypreParMatrix& edge_trueedge,
+          const mfem::Vector& edge_weight_local = mfem::Vector());
+
+    Graph(const mfem::SparseMatrix& vertex_edge_local,
```
Document.

barker29 commented on src/Graph.cpp:
```diff
@@ -0,0 +1,340 @@
+    unique_ptr<mfem::HypreParMatrix> trueedge_edge(edge_trueedge.Transpose());
+    edge_trueedge_.reset(trueedge_edge->Transpose());
+
+    if (edge_weight_local)
```
`edge_weight_local` is a reference here, not a pointer, I'm not sure this will do what you expect. Maybe `if (edge_weight_local.Size() > 0)` ?

chakshinglee replied:
Good catch, what you said should be the right way.

barker29 commented on src/LocalMixedGraphSpectralTargets.cpp:
```diff
@@ -596,10 +604,10 @@ mfem::Vector** LocalMixedGraphSpectralTargets::CollectConstant(
     const mfem::Vector& constant_vect)
 {
     // Gelever uses face_trueface rather than facedof_truedof (?) (I think one of us is just labeling it wrong)
```
```suggestion
```

barker29 commented on src/LocalMixedGraphSpectralTargets.cpp:
```diff
@@ -359,65 +427,48 @@ void MixedBlockEigensystem::ComputeEdgeTraces(mfem::DenseMatrix& evects,
+        dofs[i + num_ext_loc_dofs_diag] = dofs_offd[i] + num_ext_dofs_diag;
+}
+
 void LocalMixedGraphSpectralTargets::ComputeVertexTargets(
```
You've managed to both add functionality and make this simpler and easier to read, which is impressive.

barker29 commented on src/GraphTopology.hpp:
```diff
@@ -55,8 +56,21 @@ class GraphTopology
-    GraphTopology(mfem::SparseMatrix& vertex_edge,
-                  const mfem::HypreParMatrix& edge_d_td,
+    GraphTopology(const mfem::SparseMatrix& vertex_edge,
+                  const mfem::HypreParMatrix& edge_trueedge,
```
Modify documentation above to match name change.

barker29 reviewed (changes requested):
This is a lot of work and a big improvement.

The only thing I'm really worried about is the change in errors in all the tests - I want to understand what is causing this before I approve.

chakshinglee commented:
@barker29 I understand your concern on the huge difference in the upscaling errors. I noticed this too when I change them, and I was also surprised by the results. Here is some more info:

An observation that suggests the code is working properly:
One way to make the partition in parallel run to be same as serial run would be to use 12 processors because serial run partition the global graph into 12 partitions. I observe the same results in parallel and serial runs (up to 8 digits) if this partition is used. 

The partition when number of processors = 2 is bad (uneven):
One thing I just checked is that, the sizes of the distributed graphs are very different when 2 processes (in our tests) are used. One processor has 75 vertices, while the other gets 925. 

Right now, the number of partitions locally is set to the intended number of global partitions divided by number of processors. Based on the above split (79, 925), we can expect pretty poor partition if we divide each local graph into 6 partitions.

If I set the number of local partitions to be also proportional to relative local graph size (effectively consider "coarsening factor"). Then the errors get better (about 55% for different error measures). 

To make sure it is really just the partitioning that is causing the difference, I dont mind going back to our original partitioning method -- partition global graph, and distribute according to the global partition. This is not the main purpose of the PR anyway. I changed the partitioning method because I extracted `Graph` out, and and I think it doesn't make sense to have partition as a member of `Graph` objects. We can make this change in a separate PR. What do you think?

Another thing is, as I mentioned to you earlier, once we add the "redistribution" code, we can use the global partition to redistribute in parallel run, then we should expect the same upscale errors as before (since the partitioning should be the same). This option would be another PR also.

chakshinglee committed:
change partitioning in Upscale input from const ref to const pointer, combine two constructors

chakshinglee committed:
use default value for Wblock in Upscale parameters so that some constructors can be combined

chakshinglee committed:
switch the order of Upscale input parameters so that the code in examples/testcodes are simplified

chakshinglee committed:
fix a conditional statement

chakshinglee committed:
use all capital letters for enum type DofType

barker29 committed:
Update src/LocalMixedGraphSpectralTargets.cpp

remove some comment that is no longer needed

Co-Authored-By: chakshinglee <chakshinglee@gmail.com>

chakshinglee committed:
remove a ElementMBuilder constructor that is no longer needed

chakshinglee committed:
fix file header and update the documentation of Graph

chakshinglee committed:
more updates on documentation

chakshinglee committed:
Merge branch 'refactor/upscale' of github.com:LLNL/smoothG into refactor/upscale

chakshinglee committed:
add some documentation for the new Upscale constructor

barker29 reviewed (approved):
Thanks for all the changes, especially the nice documentation.

Regarding the error issue, I'm convinced it's okay. In a future PR, let's changethe default so the number of local partitions is  proportional to local graph size (ie "coarsening factor"), and have the tests based on that.
